examples/decomposition/apply_covid.py

```diff
@@ -0,0 +1,180 @@
+#
+# Systems serology is a new technology that examines the antibodies from a patient's serum, aiming
+# to comprehensively profile the interactions between the antibodies and
+# [Fc receptors](https://en.wikipedia.org/wiki/Fc_receptor) alongside other types of immunological
+# and demographic data. Here, we will apply CP decomposition to a [COVID-19 system serology dataset]
+# (https://www.sciencedirect.com/science/article/pii/S0092867420314598).
+#
+# Details of this analysis as well as more in-depth biological implication can be found in
+# [this work](https://www.embopress.org/doi/full/10.15252/msb.202110243).
+#
+# We first import this dataset of a panel of COVID-19 patients:
+
+import numpy as np
+import pandas as pd
+from tensorly.datasets.imports import COVID19_data
+
+data = COVID19_data()
+
+##############################################################################
+# Apply CP decomposition to this dataset with Tensorly
+# -----------------------
+# Now we apply CP decomposition to this dataset.
+
+import tensorly as tl
+from tensorly.decomposition import parafac
+
+comps = np.arange(1, 7)
+CMTFfacs = [parafac(data.tensor, cc, tol=1e-10, n_iter_max=1000,
+                    linesearch=True, orthogonalise=2) for cc in comps]
+
+##############################################################################
+# To evaluate how well CP decomposition explains the variance in the dataset, we plot the percent
+# variance reconstructed (R2X) for a range of ranks.
+
+def calcR2X(tFac, tIn=None):
+    tMask = np.isfinite(tIn)
+    vTop = np.sum(np.square(tl.cp_to_tensor(tFac) * tMask - np.nan_to_num(tIn)))
+    vBottom = np.sum(np.square(np.nan_to_num(tIn)))
+    return 1.0 - vTop / vBottom
+
+import matplotlib
+from matplotlib import pyplot as plt
+
+fig1 = plt.figure()
+ax1 = fig1.add_axes([0,0,1,1])
+CMTFR2X = np.array([calcR2X(f, data.tensor) for f in CMTFfacs])
+print(CMTFR2X)
+ax1.plot(comps, CMTFR2X, "bo")
+ax1.set_ylabel("R2X")
+ax1.set_xlabel("Number of Components")
+ax1.set_xticks([x for x in comps])
+ax1.set_xticklabels([x for x in comps])
+ax1.set_ylim(0, 1)
+ax1.set_xlim(0.0, np.amax(comps) + 0.5)
+
+##############################################################################
+# Build a Logistic Regression classifier to predict severe versus deceased COVID-19 patients
+# -----------------------
+# An important function of CP decomposition here is to reduce data while keeping significant
+# information.
+
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import roc_curve, roc_auc_score
+from sklearn.model_selection import KFold
+
+tfac = CMTFfacs[1]
+
+# Ensure that factors are negative on at most one direction.
+tfac.factors[1][:, 0] *= -1
+tfac.factors[2][:, 0] *= -1
+
+##############################################################################
+# To evaluate the classifier, we plot a ROC-AUC curve
+
+def COVIDpredict(tfac):
```

Comment:
Note a huge deal but the function is currently a little obscure to read, mostly due to none obvious variable names.

Reply:
I made a few changes to this. Let me know which variable names are still obscure.

Reply:
If we do my suggestion above then we can just get rid of this function.